File: vue-im-ui/src/shared/contracts/unknown-message.test.ts
import { describe, expect, it } from "vitest";
import { unknownMessagePresentation } from "./unknown-message";

const base = { hint: "当前版本无法显示这条消息", unsupportedText: "不支持的消息类型" };

describe("unknownMessagePresentation", () => {
  it("prefers the sender's fallback summary as the body", () => {
    const p = unknownMessagePresentation({ ...base, summary: "[投票] 周会时间", contentType: "flare.poll.v2" });
    expect(p.body).toBe("[投票] 周会时间");
    expect(p.hasSummary).toBe(true);
  });

  it("falls back to the generic hint when the summary is blank", () => {
    const p = unknownMessagePresentation({ ...base, summary: "   ", contentType: "flare.poll.v2" });
    expect(p.body).toBe("当前版本无法显示这条消息");
    expect(p.hasSummary).toBe(false);
  });

  it("uses the host label as the title when it knows the type", () => {
    expect(unknownMessagePresentation({ ...base, label: "投票" }).title).toBe("投票");
  });

  it("uses the generic unsupported wording when there is no label", () => {
    expect(unknownMessagePresentation({ ...base, label: "  " }).title).toBe("不支持的消息类型");
  });

  it("keeps the raw content type as a diagnostic, never as the body", () => {
    const p = unknownMessagePresentation({ ...base, contentType: "flare.poll.v2" });
    expect(p.diagnostic).toBe("flare.poll.v2");
    expect(p.body).toBe("当前版本无法显示这条消息");
  });

  it("reports no diagnostic when the content type is missing or blank", () => {
    expect(unknownMessagePresentation({ ...base }).diagnostic).toBe("");
    expect(unknownMessagePresentation({ ...base, contentType: "  " }).diagnostic).toBe("");
  });

  it("trims every input so stray whitespace cannot fake a value", () => {
    const p = unknownMessagePresentation({ ...base, label: " 投票 ", summary: " hi ", contentType: " x.y " });
    expect([p.title, p.body, p.diagnostic]).toEqual(["投票", "hi", "x.y"]);
  });
});

File: flutter-im-ui/test/unknown_message_test.dart
import 'package:flare_im_ui/src/components/flare_unknown_message.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

const hint = '当前版本无法显示这条消息';
const unsupported = '不支持的消息类型';

Widget _host(Widget child) =>
    MaterialApp(home: Scaffold(body: Center(child: child)));

void main() {
  test('summary wins as the body', () {
    final p = unknownMessagePresentation(
        contentType: 'flare.poll.v2',
        summary: '[投票] 周会时间',
        hint: hint,
        unsupportedText: unsupported);
    expect(p.body, '[投票] 周会时间');
    expect(p.hasSummary, isTrue);
  });

  test('blank summary falls back to the hint', () {
    final p = unknownMessagePresentation(
        summary: '   ', hint: hint, unsupportedText: unsupported);
    expect(p.body, hint);
    expect(p.hasSummary, isFalse);
  });

  test('title uses the label, else the generic wording', () {
    expect(
        unknownMessagePresentation(label: '投票', hint: hint, unsupportedText: unsupported).title,
        '投票');
    expect(
        unknownMessagePresentation(label: '  ', hint: hint, unsupportedText: unsupported).title,
        unsupported);
  });

  test('raw type stays diagnostic, never the body', () {
    final p = unknownMessagePresentation(
        contentType: 'flare.poll.v2', hint: hint, unsupportedText: unsupported);
    expect(p.diagnostic, 'flare.poll.v2');
    expect(p.body, hint);
  });

  test('missing or blank type yields no diagnostic', () {
    expect(unknownMessagePresentation(hint: hint, unsupportedText: unsupported).diagnostic, '');
    expect(
        unknownMessagePresentation(contentType: '  ', hint: hint, unsupportedText: unsupported)
            .diagnostic,
        '');
  });

  test('every input is trimmed', () {
    final p = unknownMessagePresentation(
        contentType: ' x.y ',
        label: ' 投票 ',
        summary: ' hi ',
        hint: hint,
        unsupportedText: unsupported);
    expect([p.title, p.body, p.diagnostic], ['投票', 'hi', 'x.y']);
  });

  testWidgets('renders the human placeholder and the type as diagnostic',
      (tester) async {
    await tester.pumpWidget(
        _host(const FlareUnknownMessage(contentType: 'flare.poll.v2')));
    expect(find.text(unsupported), findsOneWidget);
    expect(find.text(hint), findsOneWidget);
    expect(find.text('flare.poll.v2'), findsOneWidget);
  });

  testWidgets('offers no action without a handler', (tester) async {
    await tester.pumpWidget(_host(
        const FlareUnknownMessage(contentType: 'x.y', actionText: '了解详情')));
    expect(find.text('了解详情'), findsNothing);
  });

  testWidgets('dispatches the host action when handled', (tester) async {
    var tapped = false;
    await tester.pumpWidget(_host(FlareUnknownMessage(
        contentType: 'x.y', actionText: '了解详情', onAction: () => tapped = true)));
    await tester.tap(find.text('了解详情'));
    expect(tapped, isTrue);
  });
}
